packages/react/src/hooks/use-penumbra-transport.ts

```diff
@@ -0,0 +1,75 @@
+import {
+  type ChannelTransportOptions,
+  createChannelTransport,
+} from '@penumbra-zone/transport-dom/create';
+import { useEffect, useMemo, useState } from 'react';
+import { usePenumbra } from './use-penumbra.js';
+import { PenumbraInjectionState } from '@penumbra-zone/client';
+
+/** Unconditionally returns a Transport to the provided Penumbra context. This
+ * transport will always create synchronously, but may reject all requests if
+ * the Penumbra context does not provide a port within your configured
+ * defaultTimeoutMs (defaults to 10 seconds). */
+export const usePenumbraTransportSync = (opts?: Omit<ChannelTransportOptions, 'getPort'>) => {
+  const penumbra = usePenumbra();
+  const { port, failure, state } = penumbra;
+
+  // use a local promise to avoid re-rendering when the port is set
+  const [{ resolve: resolvePort, reject: rejectPort, promise: portPromise }] = useState(
+    Promise.withResolvers<MessagePort>(),
+  );
+
+  // memoize the transport to avoid re-creating it on every render
+  const transport = useMemo(
+    () => createChannelTransport({ ...opts, getPort: () => portPromise }),
+    [portPromise],
+  );
+
+  // handle context updates
+  useEffect(() => {
+    if (port) {
+      resolvePort(port);
+    } else if (failure) {
+      rejectPort(failure);
+    }
+  }, [failure, penumbra, port, resolvePort, rejectPort, state]);
+
+  return transport;
```

Comment:
Does the transport automatically queue up requests until the connection is ready? Or how is the queueing handled?

Reply:
behavior is in https://github.com/penumbra-zone/web/blob/main/packages/transport-dom/src/create.ts

each request awaits a shared connection attempt as part of its lifecycle, involving the `getPort` option. the connection respects the timeout configuration passed into the transport.

the difference between the two methods is that

- `usePenumbraTransportSync` will create a transport, and the transport will await a port with its timeout clock beginning at the first request.
- `usePenumbraTransport` will promise a transport, but won't init the transport until a port appears in context. the transport will spend no timeout clock.